src/ui/GraphModel.cpp

```diff
@@ -0,0 +1,349 @@
+
+    } else if (message.endpoint == block::kStagedSetting) {
+        // serviceNames is used for block's unique name in settings messages
+        fmt::print("\u001b[32mStaged settings: {}\n\u001b[0m", data);
+        handleBlockSettingsStaged(message.serviceName, data);
+
+    } else if (message.endpoint == graph::kEdgeEmplaced) {
+        handleEdgeEmplaced(data);
+
+    } else if (message.endpoint == graph::kEdgeRemoved) {
+        handleEdgeRemoved(data);
+
+    } else if (message.endpoint == graph::kGraphInspected) {
+        handleGraphRedefined(data);
+
+    } else if (message.endpoint == "LifecycleState") {
+        fmt::print("\u001b[32mNew state {}\n\u001b[0m", data);
+
+    } else {
+        fmt::print("\u001b[32mNot processed: {} data: {}\n\u001b[0m", message.endpoint, message.data.error());
+        if (!message.data) {
+            auto msg = fmt::format("Error: {}\n", message.data.error().message);
+            DigitizerUi::components::Notification::error(msg);
+        }
+    }
+    //
+}
+
+void UiGraphModel::requestGraphUpdate() {
+    gr::Message message;
+    message.cmd      = gr::message::Command::Set;
+    message.endpoint = gr::graph::property::kGraphInspect;
+    message.data     = gr::property_map{};
+    App::instance().sendMessage(message);
+}
+
+auto UiGraphModel::findBlockByName(const std::string& uniqueName) {
+    auto it = std::ranges::find_if(m_blocks, [&](const auto& block) { return block.blockUniqueName == uniqueName; });
+    return std::make_pair(it, it != m_blocks.end());
+}
+
+auto UiGraphModel::findPortByName(auto& ports, const std::string& portName) {
+    auto it = std::ranges::find_if(ports, [&](const auto& port) { return port.portName == portName; });
+    return std::make_pair(it, it != ports.end());
+}
+
+bool UiGraphModel::handleBlockRemoved(const std::string& uniqueName) {
+    auto [blockIt, found] = findBlockByName(uniqueName);
+    if (!found) {
+        // TODO warning
+        fmt::print("\u001b[32mWARNING: Unknown block removed {}\n\u001b[0m", uniqueName);
+        return false;
+    }
+
+    // Delete edges for the removed block
+    removeEdgesForBlock(*blockIt);
+    m_blocks.erase(blockIt);
+    return true;
+}
+
+void UiGraphModel::handleBlockEmplaced(const gr::property_map& blockData) {
+    const auto uniqueName  = getProperty<std::string>(blockData, "uniqueName"s);
+    const auto [it, found] = findBlockByName(uniqueName);
+    if (found) {
+        setBlockData(*it, blockData);
+    } else {
+        auto& newBlock = m_blocks.emplace_back(/*owner*/ this);
+        setBlockData(newBlock, blockData);
+    }
+}
+
+void UiGraphModel::handleBlockDataUpdated(const std::string& uniqueName, const gr::property_map& blockData) {
+    auto [blockIt, found] = findBlockByName(uniqueName);
+    if (!found) {
+        // TODO warning
+        fmt::print("\u001b[32mWARNING: Data updated for an unknown block {}\n\u001b[0m", uniqueName);
```

Comment:
Logging to the console isn't good because this will get largely lost and obfuscated other debugging efforts.
 
Could this be converted to either 
 * the new UI popup warning, or
 * message that is propagated and intercepted by the top-level scheduler?